enhancements/single-node/aos-sno-pair.md

```diff
@@ -0,0 +1,143 @@
+---
+title: aos-sno-pair
+authors:
+- @mshitrit
+  reviewers:
+- TBD
+  approvers:
+- TBD
+
+creation-date: 2021-07-12
+last-updated: 2021-07-12
+status: implementable
+---
+
+# Atomic OpenShift Single-Node-Pair
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+This enhancement describes setting up a two nodes cluster with a reduced footprint which will serve edge cases.
+
+## Motivation
+
+Some companies have a need for a highly available container management solution that fits within a reduced footprint.
+- The hardware savings are significant for customers deploying many remote sites (kiosks, branch offices, restaurant chains, etc), most notably for edge computing and RAN specifically.
+- The physical constraints of some deployments prevents more than two nodes (planes, submarines, satellites, and also RAN).
```

Comment:
The foundation of Kubernetes is etcd, which is built on extensive proven research into the Raft consensus protocol, and has lots of domain knowledge, tooling etc.  The fact that all cluster types today use etcd means we have symmetry across many cloud and bare metal cases.

I can certainly see that going from 2 to 3 nodes may increase cost for some of these users.  But it's also an enormous cost to OpenShift to add a new type of "HA" cluster and I don't think we can gloss over that at all.   I understand pacemaker etc. are also well understood and deployed in many scenarios, but those are (AFAIK) today mostly bare metal.  So we will lose out on the synergy we get today from making cloud/metal strongly aligned.

Reply:
Everywhere else that I've seen this discussed we talked about it being a thing that sits on top of OpenShift and provides HA for an app. It's aware of the OpenShift API, and manages a Deployment, but it doesn't provide HA for OpenShift itself and doesn't rely on etcd quorum.

However, some parts of this document make me think that either I'm wrong and this is about 2-node etcd HA, or we did originally explore 2-node etcd HA using this approach but abandoned it. @mshitrit could you scan through for references to etcd and make sure they're relevant and not misleading about what is being made highly available?

Reply:
>but it doesn't provide HA for OpenShift itself 

I'm not completely sure what you mean by `OpenShift itself`.
If you mean that each Single Node OpenShift (SNO) that compromise the  all-in-one OpenShift cluster (AOS) isn't HA by itself you are correct.
However the AOS as an entity is HA.

Reply:
> > but it doesn't provide HA for OpenShift itself
> 
> I'm not completely sure what you mean by `OpenShift itself`.
> If you mean that each Single Node OpenShift (SNO) that compromise the all-in-one OpenShift cluster (AOS) isn't HA by itself you are correct.
> However the AOS as an entity is HA.

Maybe the confusion is coming from the way we're using terminology around HA and the name we're giving this enhancement. This isn't about making the platform itself highly available, it's about making the application highly available.

Normally when we want an OpenShift cluster to run in a HA configuration, we say it needs at least 3 nodes. That provides for multiple replicas of etcd, the API server, etc. and makes the _platform_ highly available.

This proposal does not do that. It is about ensuring that an _application_ running on top of 2 completely distinct OpenShift instances ends up in a highly available configuration. However, those OpenShift instances are not highly available, either individually or as a pair. Someone trying to configure the instances via their APIs has to go directly to one and then the other to change both. If the etcd database or API server fails on one, access won't be redirected to the other.

I think we should call this enhancement something that more accurately describes the layer where the implementation is operating, and not "single-node pair". Using the word application in the name would help.

This design doc also talks about etcd in a way that makes the distinction between platform and application HA confusing.

For example, line 99 mentions losing etcd quorum as a risk. But etcd won't be HA, we will have 2 separate etcd instances and they don't need to have a combined quorum. 

The alternative on line 139 also mentions etcd. I don't know what "etcd learners" are or how that would relate to running HA between 2 deployments. Maybe in that case we just need a little more detail?

Reply:
learner design doc for ref: https://etcd.io/docs/v3.5/learning/design-learner/

Reply:
@cgwalters I've updated the summary to better indicate that we're not proposing messing with OCP's use of etcd.

Reply:
@hexfusion Yep, I'm still banging on about this :-)  

Learner is cool, but won't solve a 2-node network split scenario on its own.  Strictly speaking both sides should loose quorum and the cluster would be hosed.  If a winner is pre-ordained (eg. the leader), there is a decent chance that's the one that will fail and the cluster is still hosed.  If both sides are forced to retain quorum, you still need a mechanism to prevent volumes being mounted, and StatefulSets from being started, on both sides.

Reply:
done